File: java/src/main/java/br/com/emendes/problems/ContainerWithMostWater.java
package br.com.emendes.problems;

/**
 * Você recebe uma altura de matriz inteira de comprimento n.
 * Existem n linhas verticais desenhadas de modo que as duas extremidades da i-ésima linha sejam (i, 0) e (i, altura[i]).
 * <br><br>
 * Encontre duas linhas que, junto com o eixo x, formem um recipiente, de modo que o recipiente contenha mais água.
 */
public class ContainerWithMostWater {

  // Solução usando a segunda dica do problema.
  // DICA: Tente usar dois ponteiros. Defina um ponteiro para a esquerda e outro para a direita da matriz.
  // Sempre mova o ponteiro que aponta para a linha inferior.
  //
  // Solução em O(n).
  public int maxArea(int[] height) {
    int leftIndex = 0;
    int rightIndex = height.length - 1;
    int currentArea;
    int answer = 0;

    while (leftIndex < rightIndex) {
      if (height[leftIndex] < height[rightIndex]) {
        currentArea = height[leftIndex] * (rightIndex - leftIndex);
        leftIndex++;
      } else {
        currentArea = height[rightIndex] * (rightIndex - leftIndex);
        rightIndex--;
      }
      if (answer < currentArea) {
        answer = currentArea;
      }
    }

    return answer;
  }

  // Solução de força bruta, acredito que a complexida de tempo seja O(n²).
  // Como esperado, essa solução recebeu Time Limit Exceeded.
  public int badSolution(int[] height) {
    int answer = 0;
    int currentArea;

    int arrayLength = height.length;

    for (int i = 0; i < arrayLength - 1; i++) {
      for (int j = i + 1; j < arrayLength; j++) {
        if (height[i] < height[j]) {
          currentArea = height[i] * (j - i);
        } else {
          currentArea = height[j] * (j - i);
        }
        if (answer < currentArea) {
          answer = currentArea;
        }
      }
    }
    return answer;
  }

}

File: javascript/problems/ContainerWithMostWater.js
/**
 * Você recebe um array de inteiros height de tamanho n.
 * Existem n linhas verticais desenhadas de modo que os dois pontos finais da i-ésima linha são
 * (i, 0) e (i, altura[i]).
 *
 * Encontre duas retas que, juntamente com o eixo x, formem um recipiente, de modo que o recipiente
 * contenha mais água.
 *
 * Retorne a quantidade máxima de água que um recipiente pode armazenar.
 *
 * Observe que você não pode inclinar o recipiente.
 *
 * Restrições:
 *
 * n == height.length
 *
 * 2 <= n <= 10⁵
 *
 * 0 <= height[i] <= 10⁴
 *
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
  let left = 0;
  let right = height.length - 1;
  let maxArea = 0;

  while (left < right) {
    let shorterHeight = Math.min(height[left], height[right]);
    maxArea = Math.max(maxArea, shorterHeight * (right - left));

    if (height[left] < height[right]) left++;
    else right--;
  }

  return maxArea;
};
